File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Models/Receipt.swift
//
//  Receipt.swift
//  PaykitDemo
//
//  Receipt model for payment history tracking.
//

import Foundation

/// Payment status
public enum PaymentReceiptStatus: String, Codable {
    case pending = "pending"
    case completed = "completed"
    case failed = "failed"
    case refunded = "refunded"
}

/// Payment direction
public enum PaymentDirection: String, Codable {
    case sent = "sent"
    case received = "received"
}

/// A payment receipt (local model, different from PaykitMobile.Receipt)
public struct PaymentReceipt: Identifiable, Codable, Equatable {
    /// Unique identifier
    public let id: String
    /// Direction of payment
    public let direction: PaymentDirection
    /// Counterparty public key (z-base32)
    public let counterpartyKey: String
    /// Counterparty display name (if known)
    public var counterpartyName: String?
    /// Amount in satoshis
    public let amountSats: UInt64
    /// Payment status
    public var status: PaymentReceiptStatus
    /// Payment method used
    public let paymentMethod: String
    /// When the payment was initiated
    public let createdAt: Date
    /// When the payment was completed (if applicable)
    public var completedAt: Date?
    /// Optional memo/note
    public var memo: String?
    /// Transaction ID (if applicable)
    public var txId: String?
    /// Payment proof (optional, JSON string)
    public var proof: String?
    /// Whether proof has been verified
    public var proofVerified: Bool = false
    /// Timestamp when proof was verified
    public var proofVerifiedAt: Date?
    
    public init(
        direction: PaymentDirection,
        counterpartyKey: String,
        counterpartyName: String? = nil,
        amountSats: UInt64,
        paymentMethod: String,
        memo: String? = nil
    ) {
        self.id = UUID().uuidString
        self.direction = direction
        self.counterpartyKey = counterpartyKey
        self.counterpartyName = counterpartyName
        self.amountSats = amountSats
        self.status = .pending
        self.paymentMethod = paymentMethod
        self.createdAt = Date()
        self.completedAt = nil
        self.memo = memo
        self.txId = nil
        self.proof = nil
        self.proofVerified = false
        self.proofVerifiedAt = nil
    }
    
    /// Mark as completed
    mutating func complete(txId: String? = nil) {
        self.status = .completed
        self.completedAt = Date()
        self.txId = txId
    }
    
    /// Mark as failed
    mutating func fail() {
        self.status = .failed
    }
}

extension PaymentReceipt {
    /// Abbreviated counterparty key for display
    var abbreviatedCounterparty: String {
        guard counterpartyKey.count > 16 else { return counterpartyKey }
        let prefix = counterpartyKey.prefix(8)
        let suffix = counterpartyKey.suffix(8)
        return "\(prefix)...\(suffix)"
    }
    
    /// Display name (name if known, otherwise abbreviated key)
    var displayName: String {
        counterpartyName ?? abbreviatedCounterparty
    }
    
    /// Formatted amount
    var formattedAmount: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        return "\(direction == .sent ? "-" : "+")\(formatter.string(from: NSNumber(value: amountSats)) ?? "\(amountSats)") sats"
    }
    
    /// Status color name (for SwiftUI)
    var statusColorName: String {
        switch status {
        case .pending: return "orange"
        case .completed: return "green"
        case .failed: return "red"
        case .refunded: return "purple"
        }
    }
    
    /// Create a local PaymentReceipt from an FFI Receipt
    /// - Parameters:
    ///   - ffiReceipt: The FFI Receipt from PaykitClient.createReceipt()
    ///   - direction: Whether this is a sent or received payment
    ///   - counterpartyName: Optional display name for the counterparty
    /// - Returns: A local PaymentReceipt for storage
    static func fromFFI(
        _ ffiReceipt: Receipt,
        direction: PaymentDirection,
        counterpartyName: String? = nil
    ) -> PaymentReceipt {
        let counterpartyKey = direction == .sent ? ffiReceipt.payee : ffiReceipt.payer
        let amountSats = UInt64(ffiReceipt.amount ?? "0") ?? 0
        
        var receipt = PaymentReceipt(
            direction: direction,
            counterpartyKey: counterpartyKey,
            counterpartyName: counterpartyName,
            amountSats: amountSats,
            paymentMethod: ffiReceipt.methodId
        )
        
        // Override the auto-generated ID with the FFI receipt ID
        // We need to use a different initializer approach
        return PaymentReceipt(
            id: ffiReceipt.receiptId,
            direction: direction,
            counterpartyKey: counterpartyKey,
            counterpartyName: counterpartyName,
            amountSats: amountSats,
            status: .pending,
            paymentMethod: ffiReceipt.methodId,
            createdAt: Date(timeIntervalSince1970: Double(ffiReceipt.createdAt)),
            completedAt: nil,
            memo: nil,
            txId: nil,
            proof: nil,
            proofVerified: false,
            proofVerifiedAt: nil
        )
    }
    
    /// Mark proof as verified
    mutating func markProofVerified() {
        self.proofVerified = true
        self.proofVerifiedAt = Date()
    }
}

// MARK: - Full Initializer for FFI Integration

extension PaymentReceipt {
    /// Full initializer for creating receipts with all fields
    init(
        id: String,
        direction: PaymentDirection,
        counterpartyKey: String,
        counterpartyName: String?,
        amountSats: UInt64,
        status: PaymentReceiptStatus,
        paymentMethod: String,
        createdAt: Date,
        completedAt: Date?,
        memo: String?,
        txId: String?,
        proof: String?,
        proofVerified: Bool,
        proofVerifiedAt: Date?
    ) {
        self.id = id
        self.direction = direction
        self.counterpartyKey = counterpartyKey
        self.counterpartyName = counterpartyName
        self.amountSats = amountSats
        self.status = status
        self.paymentMethod = paymentMethod
        self.createdAt = createdAt
        self.completedAt = completedAt
        self.memo = memo
        self.txId = txId
        self.proof = proof
        self.proofVerified = proofVerified
        self.proofVerifiedAt = proofVerifiedAt
    }
}

File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/model/Receipt.kt
package com.paykit.demo.model

import kotlinx.serialization.Serializable
import java.util.UUID

/**
 * Payment status
 */
@Serializable
enum class PaymentStatus {
    PENDING,
    COMPLETED,
    FAILED,
    REFUNDED
}

/**
 * Payment direction
 */
@Serializable
enum class PaymentDirection {
    SENT,
    RECEIVED
}

/**
 * A payment receipt
 */
@Serializable
data class Receipt(
    /** Unique identifier */
    val id: String,
    /** Direction of payment */
    val direction: PaymentDirection,
    /** Counterparty public key (z-base32) */
    val counterpartyKey: String,
    /** Counterparty display name (if known) */
    var counterpartyName: String? = null,
    /** Amount in satoshis */
    val amountSats: Long,
    /** Payment status */
    var status: PaymentStatus = PaymentStatus.PENDING,
    /** Payment method used */
    val paymentMethod: String,
    /** When the payment was initiated (Unix timestamp) */
    val createdAt: Long = System.currentTimeMillis(),
    /** When the payment was completed (Unix timestamp, if applicable) */
    var completedAt: Long? = null,
    /** Optional memo/note */
    var memo: String? = null,
    /** Transaction ID (if applicable) */
    var txId: String? = null,
    /** Payment proof (optional, JSON string) */
    var proof: String? = null,
    /** Whether proof has been verified */
    var proofVerified: Boolean = false,
    /** Timestamp when proof was verified (Unix timestamp in milliseconds) */
    var proofVerifiedAt: Long? = null
) {
    companion object {
        /**
         * Create a new receipt
         */
        fun create(
            direction: PaymentDirection,
            counterpartyKey: String,
            counterpartyName: String? = null,
            amountSats: Long,
            paymentMethod: String,
            memo: String? = null
        ): Receipt {
            return Receipt(
                id = UUID.randomUUID().toString(),
                direction = direction,
                counterpartyKey = counterpartyKey,
                counterpartyName = counterpartyName,
                amountSats = amountSats,
                paymentMethod = paymentMethod,
                memo = memo
            )
        }
        
        /**
         * Create a local Receipt from an FFI Receipt
         * @param ffiReceipt The FFI Receipt from PaykitClient.createReceipt()
         * @param direction Whether this is a sent or received payment
         * @param counterpartyName Optional display name for the counterparty
         * @return A local Receipt for storage
         */
        fun fromFFI(
            ffiReceipt: com.paykit.mobile.Receipt,
            direction: PaymentDirection,
            counterpartyName: String? = null
        ): Receipt {
            val counterpartyKey = if (direction == PaymentDirection.SENT) {
                ffiReceipt.payee
            } else {
                ffiReceipt.payer
            }
            val amountSats = ffiReceipt.amount?.toLongOrNull() ?: 0L
            
            return Receipt(
                id = ffiReceipt.receiptId,
                direction = direction,
                counterpartyKey = counterpartyKey,
                counterpartyName = counterpartyName,
                amountSats = amountSats,
                paymentMethod = ffiReceipt.methodId,
                createdAt = ffiReceipt.createdAt * 1000 // Convert to milliseconds
            )
        }
    }
    
    /**
     * Mark proof as verified
     */
    fun markProofVerified(): Receipt {
        return this.copy(
            proofVerified = true,
            proofVerifiedAt = System.currentTimeMillis()
        )
    }
    
    /**
     * Mark as completed
     */
    fun complete(txId: String? = null): Receipt {
        return copy(
            status = PaymentStatus.COMPLETED,
            completedAt = System.currentTimeMillis(),
            txId = txId
        )
    }
    
    /**
     * Mark as failed
     */
    fun fail(): Receipt {
        return copy(status = PaymentStatus.FAILED)
    }
    
    /**
     * Abbreviated counterparty key for display
     */
    val abbreviatedCounterparty: String
        get() {
            if (counterpartyKey.length <= 16) return counterpartyKey
            return "${counterpartyKey.take(8)}...${counterpartyKey.takeLast(8)}"
        }
    
    /**
     * Display name (name if known, otherwise abbreviated key)
     */
    val displayName: String
        get() = counterpartyName ?: abbreviatedCounterparty
    
    /**
     * Formatted amount with direction sign
     */
    val formattedAmount: String
        get() {
            val sign = if (direction == PaymentDirection.SENT) "-" else "+"
            return "$sign${String.format("%,d", amountSats)} sats"
        }
}

